core/state/statedb.go

```diff
@@ -623,19 +623,40 @@ func (s *StateDB) GetOrNewStateObject(addr common.Address) *stateObject {
 // the given address, it is overwritten and returned as the second return value.
 func (s *StateDB) createObject(addr common.Address) (newobj, prev *stateObject) {
 	prev = s.getDeletedStateObject(addr) // Note, prev might have been deleted, we need that!
-
-	var prevdestruct bool
-	if prev != nil {
-		_, prevdestruct = s.stateObjectsDestruct[prev.address]
-		if !prevdestruct {
-			s.stateObjectsDestruct[prev.address] = struct{}{}
-		}
-	}
 	newobj = newObject(s, addr, types.StateAccount{})
 	if prev == nil {
 		s.journal.append(createObjectChange{account: &addr})
 	} else {
-		s.journal.append(resetObjectChange{prev: prev, prevdestruct: prevdestruct})
+		// The original account should be marked as destructed and all cached
+		// account and storage data should be cleared as well. Note, it must
+		// be done here, otherwise the destruction event of original one will
+		// be lost.
+		_, prevdestruct := s.stateObjectsDestruct[prev.address]
+		if !prevdestruct {
+			s.stateObjectsDestruct[prev.address] = struct{}{}
+		}
+		var (
+			account []byte
+			storage map[common.Hash][]byte
+		)
+		// There may be some cached account/storage data already since IntermediateRoot
+		// will be called for each transaction before byzantium fork which will always
+		// cache the latest account/storage data.
+		if s.snapAccounts != nil {
+			account = s.snapAccounts[prev.addrHash]
+			delete(s.snapAccounts, prev.addrHash)
+		}
+		if s.snapStorage != nil {
+			storage = s.snapStorage[prev.addrHash]
+			delete(s.snapStorage, prev.addrHash)
+		}
```

Comment:
These storages are otherwise only manipulated at transaction boundaries: finalize and commit. It does not seem correct to me to modify them on the fly like this, in the middle of a transaction.

What I most of all want to ensure that we avoid: changing the evm semantics in a way which fixes a theoretical corner-case but causes a break in something which can be triggered in practice.

Reply:
`snapAccount`, `snapStorage` are essentially equivalent with `stateObjectsDestruct`.
- `snapAccount`, `snapStorage` cache the state data
- `stateObjectsDestruct` caches the destruction event

Because in the case of `ResetAccount`, the original one is replaced with the new one,
it's the only correct place to manipulate these three maps, otherwise the destruction
event will be lost.

We can run a full sync to ensure nothing is broken.